### ISSUE TITLE
Unity port

### ISSUE THREAD

**Issue by KHCmaster** [KHCmaster/PPD#7]
## Why?

Current PPD game framework is written 10 years ago by me.
There was no good game engine on which I can use C#.
But now, there is Unity. Unity can support 2D/3D scene, C# scripting, cross platform, etc.

## How about compatibility?

I don't care about compatibility between 2.0 and 1.x.
Breaking change about setting, style, input, sound, movie and mod will happen.

## First step

I don't have any unity experience.
So if someone who has it, please create sample project like PPD.

Feature

- [ ] Create Unity folder besides Win and create project.
- [ ] Load *.ppd by PPDStructure
- [ ] Create Mark(PPDCore) and move it by time. No sound, no movie is allowed. Just consider it passes 1/60 sec in each frame.
- [ ] Load image from resources(static embed assets is allowed).

**Comment by tmillerklugman** [comment]
Hi, I saw this and just wanted to chime in that unity is a horrible engine to use for rhythm games, in fact with the built in audio engine there would be  be 100ms+ latency. I would recommend taking a look at Osu Framework. Which is built for making rhythm games and you could maybe in the future do a pull request to add a diva mode directly into osu with ranked leader boards and local scores for csinput. Which is my biggest gripe at the moment. I love saving my scores and not being able to do that with the DS4 controller is very unfortunate. I will drop the link to Osu!Framework. https://github.com/ppy/osu-framework (as a bonus it is cross platform thanks to .net core and uses c#) Hope this helps out in making PPD a much more stable experience!

**Comment by EIREXE** [comment]
I would like to suggest Godot too, it's what I use for my PD style game and it has features with rhythm games in mind:

https://docs.godotengine.org/en/latest/tutorials/audio/sync_with_audio.html

**Comment by dogtopus** [comment]
> Hi, I saw this and just wanted to chime in that unity is a horrible engine to use for rhythm games, in fact with the built in audio engine there would be be 100ms+ latency.

@Dengekiko Hm that somewhat contradicts to the fact that a lot of rhythm games and most of SEGA's current gen arcade rhythm games are based on Unity. If the audio engine is horrible how would they workaround this? I never reverse engineered any of them that deep so I don't really know.

**Comment by EIREXE** [comment]
> > Hi, I saw this and just wanted to chime in that unity is a horrible engine to use for rhythm games, in fact with the built in audio engine there would be be 100ms+ latency.
> 
> @Dengekiko Hm that somewhat contradicts to the fact that a lot of rhythm games and most of SEGA's current gen arcade rhythm games are based on Unity. If the audio engine is horrible how would they workaround this? I never reverse engineered any of them that deep so I don't really know.
> 
> As a half freetard I would suggest Godot too since Unity is not open-source and osu framework is not fully open source. OTOH seems that Godot also supports C# and there are documented audio latency compensation APIs as previously mentioned here.

What most people do is write native code that replaced the input and audio playback stack in unity.

The only thing godot lacks is a timestamped (and possibly multi threaded) input stack.

My diva clone however has been said to have similar input latency to AFT, so I suspect that might be unecessary if the performance is good enough.

**Comment by dogtopus** [comment]
Just checked out your game (Project Heartbeat) and it looks pretty good. Will it be open sourced eventually after it's done? I (and probably some others who love open source and Linux) would be really glad to have an open source and cross platform Project Diva simulator.

**Comment by EIREXE** [comment]
> 
> 
> Just checked out your game (Project Heartbeat) and it looks pretty good. Will it be open sourced eventually after it's done? I (and probably some others who love open source and Linux) would be really glad to have an open source and cross platform Project Diva simulator.

Yes, it will be open sourced a few months after it goes out of early access.
